### PR TITLE
235 / Configurable encoding of events to/from bytes

### LINKED ISSUE #83: [project] Use JSON as encoding for most data
To use a more standard encoding format for events, commands etc.

### LINKED ISSUE #235: Configurable encoding of events to/from bytes
I created a package to remove duplicate logic when dealing with encoding/decoding `eh.Event`. This issue is about of integration.

Repo: https://github.com/giautm/eh-encoder

Example with eventbus:
```go
import (
	"github.com/giautm/eh-encoder"
	"github.com/giautm/eh-encoder/bson"
)

type EventBus struct {
	// ...   
	encoder      encoder.Encoder
}

// NewEventBus creates a EventBus.
func NewEventBus(
) (*EventBus, error) {
	return &EventBus{
		// ...   
		encoder:            bson.NewEncoder(), // or json.NewEncoder()
	}, nil
}

// PublishEvent publishes an event to all handlers capable of handling it.
func (b *EventBus) PublishEvent(ctx context.Context, event eh.Event) error {
	data, err := b.encoder.Encode(ctx, event)
	if err != nil {
		return err
	}

	_, _, err = b.producer.SendMessage(&sarama.ProducerMessage{
		Topic: b.producerTopicFunc(event),
		Value: sarama.StringEncoder(data),
	})
	if err != nil {
		return ErrCouldNotPublishEvent(err)
	}

	return nil
}
```

### PR DESCRIPTION
**Pull request by maxekman**
### Description

Add an EventCodec interface which is used in event busses.

### Affected Components

- New component: Codec
- Event Bus

### Related Issues

Closes #235, closes #83

### CODE REVIEW

**Comment by maxekman** [comment]
@mvmaasakkers Ready for review, if you want.

File: codec/acceptance_testing.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// Codec should pass. It should manually be called from a test case in each
+// implementation:
+//
+//   func TestEventCodec(t *testing.T) {
+//       c := EventCodec{}
+//       expectedBytes = []byte("")
+//       eventbus.AcceptanceTest(t, c, expectedBytes)
+//   }
+//
+func EventCodecAcceptanceTest(t *testing.T, c eh.EventCodec, expectedBytes []byte) {
```
Comment (mvmaasakkers) [inline codec/acceptance_testing.go:38]: Maybe it would be good to add some more tests with for example nested structs, nil definitions, single bools etc.

File: codec/json/codec.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+
+	eh "github.com/looplab/eventhorizon"
+)
+
+// EventCodec is a codec for marshaling and unmarshaling events
+// to and from bytes in BSON format.
```
Comment (mvmaasakkers) [inline codec/json/codec.go:30]: This file needs a global string replace of BSON -> JSON

**Review by mvmaasakkers — CHANGES_REQUESTED** [review]

File: codec/acceptance_testing.go (hunk 1 above)
Comment (maxekman) [reply codec/acceptance_testing.go:38]: Sounds good!

File: codec/json/codec.go (hunk 2 above)
Comment (maxekman) [reply codec/json/codec.go:30]: Oh, snap! 😕

**Comment by maxekman** [comment]
@mvmaasakkers Updated with your suggestions.

**Review by mvmaasakkers — APPROVED** [review]
